observer/holder/nodesHolder.go

```diff
@@ -0,0 +1,217 @@
+package holder
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	logger "github.com/multiversx/mx-chain-logger-go"
+	"github.com/multiversx/mx-chain-proxy-go/data"
+)
+
+type cacheType string
+
+const (
+	syncedNodesCache       cacheType = "syncedNodes"
+	outOfSyncNodesCache    cacheType = "outOfSyncNodes"
+	syncedFallbackCache    cacheType = "syncedFallbackNodes"
+	outOfSyncFallbackNodes cacheType = "outOfSyncFallbackNodes"
+)
+
+var (
+	log               = logger.GetOrCreate("observer/holder")
+	errEmptyNodesList = errors.New("empty nodes list")
+)
+
+type nodesHolder struct {
+	mut          sync.RWMutex
+	allNodes     map[uint32][]*data.NodeData
+	cache        map[string][]*data.NodeData
+	availability data.ObserverDataAvailabilityType
+}
+
+// NewNodesHolder will return a new instance of a nodesHolder
+func NewNodesHolder(syncedNodes []*data.NodeData, fallbackNodes []*data.NodeData, availability data.ObserverDataAvailabilityType) (*nodesHolder, error) {
+	if len(syncedNodes) == 0 && len(fallbackNodes) == 0 && availability != data.AvailabilityRecent {
+		return nil, errEmptyNodesList
+	}
+	return &nodesHolder{
+		allNodes:     computeInitialNodeList(syncedNodes, fallbackNodes),
+		cache:        make(map[string][]*data.NodeData),
+		availability: availability,
+	}, nil
+}
+
+// UpdateNodes will update the internal maps based on the provided nodes
+func (nh *nodesHolder) UpdateNodes(nodesWithSyncStatus []*data.NodeData) {
+	if len(nodesWithSyncStatus) == 0 {
+		return
+	}
+
+	nh.mut.Lock()
+	defer nh.mut.Unlock()
+
+	nh.allNodes = make(map[uint32][]*data.NodeData)
+	nh.cache = make(map[string][]*data.NodeData)
+	for _, node := range nodesWithSyncStatus {
+		shouldSkipNode := nh.availability == data.AvailabilityRecent && !node.IsSnapshotless ||
```

Comment:
please split this in 2. You have 2 `&&` operators along with one `||`

Reply:
done